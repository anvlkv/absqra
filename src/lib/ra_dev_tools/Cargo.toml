[package]
name = "ra_dev_tools"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"
build = "build.rs"

[build-dependencies]
build-deps = "0.1.4"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = "0.16.0"
regex = "1.3.9"
example_files_macro = {path="./src/lib/example_files_macro"}
files_iterator = {path="./src/lib/files_iterator"}
