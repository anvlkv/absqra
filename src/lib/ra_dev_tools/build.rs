extern crate build_deps;

fn main() {
    // Enumerate files in sub-folder "examples/*", being relevant for the test-generation (as example)
    // If function returns with error, exit with error message.
    build_deps::rerun_if_changed_paths( "../../../examples/*/*" ).unwrap();

    // Adding the parent directory "examples" to the watch-list will capture new-files being added
    build_deps::rerun_if_changed_paths( "../../../examples/*" ).unwrap();
}