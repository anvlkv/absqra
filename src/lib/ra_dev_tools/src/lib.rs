pub extern crate insta;
pub use files_iterator::for_each_ra_example_file;

#[allow(unused_attributes)]
#[macro_use] pub use example_files_macro::make_example_tests;

#[cfg(test)]
mod test {
    use super::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_works_with(content: String, file_name: String) {
        println!("didn't panic with content: \n[{}] of file_name: {}", content, file_name);
    }
}