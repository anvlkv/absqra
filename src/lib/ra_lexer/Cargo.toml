[package]
name = "ra_lexer"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
failure = "0.1.8"
failure_derive = "0.1.8"
serde = "1.0.114"
serde_json = "1.0.56"
serde_derive = "1.0.114"

[dev-dependencies]
ra_dev_tools = {path = "../ra_dev_tools"}
