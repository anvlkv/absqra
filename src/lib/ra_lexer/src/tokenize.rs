use super::*;

pub fn tokenize<'a>(input: &'a str) -> impl Iterator<Item = Result<RaToken, LexerError>> + 'a {
    Cursor::from(input)
    // tokenize_cursor(Cursor::new(input, Position(1, 0), 0, 0)).map(|lexer_result| lexer_result?.try_into())
}