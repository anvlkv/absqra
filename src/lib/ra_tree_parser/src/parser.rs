use super::*;
use ra_lexer::errors::LexerError;

pub fn parse<'a>(
    mut tokens_stream: impl Iterator<Item = Result<RaToken, LexerError>>
) -> Result<RaTree, (Vec<LexerError>, Option<RaTree>)> {

    let mut tree = RaTree::new();

    let mut errors: Vec<LexerError> = Vec::new();

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                tree.add_token(token);
            },
            Err(e) => {
                errors.push(e);
            }
        }
    }
    if errors.len() > 0 {
        Err((errors, Some(tree)))
    }
    else {
        Ok(tree)
    }
}
