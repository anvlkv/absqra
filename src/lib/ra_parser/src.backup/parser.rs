use crate::expressions::context_expression::{ContextExpression, ContextExpressionMemberKind, ContextExpressionMember};
use ra_lexer::token::{RaToken, TokenKind};
use ra_lexer::tokenize;
use ra_lexer::errors::LexerError;

use super::block::{Block, BlockKind};
use super::cursor::Cursor;
use super::errors::ParserError;
use super::expressions::traits::{Leveled, Expandable, Positioned};
use failure::Backtrace;

pub fn parse<'a>(input: &'a str) -> Result<Block<'a>, (Vec<ParserError>, Block<'a>)> {
    let stream = tokenize(input);
    
    let mut errors: Vec<ParserError> = Vec::new();

    let program = {
        let mut block = Block::new(RaToken::default()).unwrap();

        let mut tokens_cursor = Cursor::new(
            stream
            .take_while(|r| r.is_ok())
            .map(|r| r.unwrap())
            .filter(|tok| tok.kind.unwrap() != TokenKind::Comment)
        );
        loop {
            if !tokens_cursor.is_eof() {
                match tokens_cursor.advance_block() {
                    Ok(blk) => block.children.push(blk),
                    Err(es) => errors.extend(es.into_iter()),
                }
            } else {
                break;
            }
        }

        block
    };

    if errors.len() > 0 {
        Err((errors, program))
    } else {
        Ok(program)
    }
}

impl<'token, I> Cursor<'token, I>
where
    I: Iterator<Item = RaToken<'token>>,
{
    fn advance_block<'a>(&'a mut self) -> Result<Block<'token>, Vec<ParserError>> {
        let first_token = self.bump();
        let mut errors: Vec<ParserError> = Vec::new();
        match first_token {
            Some(token) => {
                self.parse_block(token)
            },
            None => {
                errors.push(ParserError::UnexpectedEndOfInput(self.position, Backtrace::new()));
                Err(errors)
            }
        }
    }

    fn parse_block(
        &mut self,
        first_token: RaToken<'token>,
    ) -> Result<Block<'token>, Vec<ParserError>> {
        let mut errors: Vec<ParserError> = Vec::new();
        let block = Block::new(first_token)?;

        if let (Some(blk), mut errs) =
            self.check_parse_block_expression(block, &mut errors)
        {
            if let (Some(blk), mut errs) = self.check_parse_block_children(blk, &mut errs) {
                if let (Some(blk), errs) = self.check_parse_union(blk, &mut errs) {
                    if errs.len() == 0 {
                        return Ok(blk);
                    }
                }
            }
        }

        Err(errors)
    }

    fn check_parse_block_expression<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut lvl = block.get_level();
        let mut line_number = (block.get_position().0).0;
        while self.first_ahead().is_some()
            && self.first_ahead().unwrap().level == lvl
            && (self.first_ahead().unwrap().position.0).0 == line_number {
                match block.clone().append_item(self.bump().unwrap()) {
                    Ok(blk) => {
                        block = blk;

                        // // TODO: ugly, move this logic elsewhere

                        if let BlockKind::ContextModification(Some(ContextExpression(target, source))) = block.clone().kind { 
                            if ContextExpressionMember::Target(ContextExpressionMemberKind::MSpecifier) == target && source.is_none() {
                                if let (Some(mut blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    let close_curly_brace = self.bump().unwrap();
                                    assert_eq!(close_curly_brace.kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    lvl = close_curly_brace.level;
                                    line_number = (close_curly_brace.position.0).0;
                            //         // let updated_expression = ContextExpression(target, None).append_item(self.bump().unwrap())?;

                            //         // blk.kind = BlockKind::ContextModification(Some(updated_expression));
                            //         // self.check_parse_block_expression(blk, errs)
                                }
                            }
                            else if source.is_some() && ContextExpressionMember::Source(ContextExpressionMemberKind::MSpecifier) == source.unwrap() {
                                if let (Some(blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    // return Ok(blk);
                                }
                            }
                            println!("{:?}", block);
                        }

                        // // end TODO
                    },
                    Err(e) => {
                        errors.push(e)
                    }
                }
        }

        (Some(block), errors)
    }

    fn check_parse_block_children<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        
        while !self.is_eof() && self.first_ahead().unwrap().level == block.get_level() + 1 {
            match self.advance_block() {
                Ok(child) => {
                    block.children.push(child);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }

        if errors.len() == 0 {
            (Some(block), errors)
        } else {
            (None, errors)
        }
    }

    fn check_parse_union<'errors>(
        &mut self,
        block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut union_block = block.clone();
        let lvl = block.get_level();
        let mut union_size = 0;
        while !self.is_eof()
            && self.first_ahead().unwrap().kind.unwrap() == TokenKind::Coma
            && self.first_ahead().unwrap().level == lvl
        {
            assert!(self.bump().unwrap().kind.unwrap() == TokenKind::Coma);
            match self.advance_block() {
                Ok(blk) => {
                    union_size += 1;
                    union_block.children.push(blk);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }
        
        if union_size > 0 {
            union_block.kind = BlockKind::Union(union_size);
            (Some(union_block), errors)
        } else {
            (Some(block), errors)
        }
    }
}
