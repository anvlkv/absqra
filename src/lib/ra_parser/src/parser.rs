use crate::errors::ParserError;
use crate::block::Block;
use crate::parsed_by_token::ParsedByToken;
use ra_lexer::tokenize;
use ra_lexer::token::TokenKind;

pub fn parse<'a>(input: &'a str) -> Result<Block, (Vec<ParserError>, Option<Block>)> {
    let mut errors: Vec<ParserError> = Vec::new();
    let mut tokens_stream = tokenize(input);
    
    let mut block = {
        match Block::new_program() {
            Ok(b) => b,
            Err(e) => {
                errors.extend(e);
                return Err((errors, None));
            }
        }
    };

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                if token.kind == TokenKind::Comment {
                    continue;
                }
                else {
                    match block.clone().append_token(token) {
                        Ok(blk) => block=blk,
                        Err(e) => errors.extend(e)
                    }
                }
            },
            Err(e) => {
                errors.push(e.into());
            }
        }
    }

    if errors.len() == 0 {
        Ok(*block)
    }
    else {
        Err((errors, Some(*block)))
    }
}