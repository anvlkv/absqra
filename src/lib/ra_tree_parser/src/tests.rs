
mod lib {
    use ra_dev_tools::insta::{assert_snapshot, assert_debug_snapshot};    
    use ra_dev_tools::{make_example_tests};
    use ra_lexer::tokenize;
    use crate::parser::parse;

    #[make_example_tests]
    #[test]
    fn it_should_match_snapshots(contents: String, file_name: String) {
        let err_messages_file_name = format!("{}__{}",file_name, "ERR__MESSAGES");
        let errors_file_name = format!("{}__{}", file_name, "ERR");
        let tree = {
            let mut tokens_iter = tokenize(&contents);
            match parse(tokens_iter) {
                Ok(tree) => {
                    tree
                },
                Err((errors, parsed)) => {
                    let formatted_errors: String = {
                        errors.into_iter().map(|e| format!("{}", e)).collect::<Vec<String>>().join("\n")
                    };
                    
                    assert_snapshot!(err_messages_file_name, formatted_errors);
                    assert_debug_snapshot!(errors_file_name, parsed);


                    panic!("failed to parse example {:?}", file_name);
                }
            }
        };
        assert_debug_snapshot!(file_name, tree)       
    }
}
