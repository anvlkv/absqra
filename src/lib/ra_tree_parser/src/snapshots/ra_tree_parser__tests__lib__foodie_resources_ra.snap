---
source: src/lib/ra_tree_parser/src/tests.rs
expression: tree
---
{
  "tokens": [],
  "level": 0,
  "position": [
    [
      0,
      0
    ],
    [
      57,
      6
    ]
  ],
  "children": [
    {
      "tokens": [
        {
          "kind": "Colon",
          "position": [
            [
              1,
              1
            ],
            [
              1,
              2
            ]
          ],
          "level": 1
        },
        {
          "kind": {
            "Identifier": "navigationItems"
          },
          "position": [
            [
              1,
              2
            ],
            [
              1,
              17
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          1,
          1
        ],
        [
          22,
          10
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": {
                "Identifier": "meals"
              },
              "position": [
                [
                  2,
                  5
                ],
                [
                  2,
                  10
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              2,
              5
            ],
            [
              8,
              10
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      3,
                      9
                    ],
                    [
                      3,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      3,
                      10
                    ],
                    [
                      3,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      3,
                      12
                    ],
                    [
                      3,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "en"
                  },
                  "position": [
                    [
                      3,
                      13
                    ],
                    [
                      3,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Meals"
                  },
                  "position": [
                    [
                      3,
                      15
                    ],
                    [
                      5,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  3,
                  9
                ],
                [
                  5,
                  10
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      6,
                      9
                    ],
                    [
                      6,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      6,
                      10
                    ],
                    [
                      6,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      6,
                      12
                    ],
                    [
                      6,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ru"
                  },
                  "position": [
                    [
                      6,
                      13
                    ],
                    [
                      6,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Блюда"
                  },
                  "position": [
                    [
                      6,
                      15
                    ],
                    [
                      8,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  6,
                  9
                ],
                [
                  8,
                  10
                ]
              ],
              "children": []
            }
          ]
        },
        {
          "tokens": [
            {
              "kind": {
                "Identifier": "countries"
              },
              "position": [
                [
                  9,
                  5
                ],
                [
                  9,
                  14
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              9,
              5
            ],
            [
              15,
              10
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      10,
                      9
                    ],
                    [
                      10,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      10,
                      10
                    ],
                    [
                      10,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      10,
                      12
                    ],
                    [
                      10,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "en"
                  },
                  "position": [
                    [
                      10,
                      13
                    ],
                    [
                      10,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Countries"
                  },
                  "position": [
                    [
                      10,
                      15
                    ],
                    [
                      12,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  10,
                  9
                ],
                [
                  12,
                  10
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      13,
                      9
                    ],
                    [
                      13,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      13,
                      10
                    ],
                    [
                      13,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      13,
                      12
                    ],
                    [
                      13,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ru"
                  },
                  "position": [
                    [
                      13,
                      13
                    ],
                    [
                      13,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Национальные кухни"
                  },
                  "position": [
                    [
                      13,
                      15
                    ],
                    [
                      15,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  13,
                  9
                ],
                [
                  15,
                  10
                ]
              ],
              "children": []
            }
          ]
        },
        {
          "tokens": [
            {
              "kind": {
                "Identifier": "restrictions"
              },
              "position": [
                [
                  16,
                  5
                ],
                [
                  16,
                  17
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              16,
              5
            ],
            [
              22,
              10
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      17,
                      9
                    ],
                    [
                      17,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      17,
                      10
                    ],
                    [
                      17,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      17,
                      12
                    ],
                    [
                      17,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ru"
                  },
                  "position": [
                    [
                      17,
                      13
                    ],
                    [
                      17,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Preferences"
                  },
                  "position": [
                    [
                      17,
                      15
                    ],
                    [
                      19,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  17,
                  9
                ],
                [
                  19,
                  10
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      20,
                      9
                    ],
                    [
                      20,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      20,
                      10
                    ],
                    [
                      20,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      20,
                      12
                    ],
                    [
                      20,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ru"
                  },
                  "position": [
                    [
                      20,
                      13
                    ],
                    [
                      20,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Предпочтения"
                  },
                  "position": [
                    [
                      20,
                      15
                    ],
                    [
                      22,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  20,
                  9
                ],
                [
                  22,
                  10
                ]
              ],
              "children": []
            }
          ]
        }
      ]
    },
    {
      "tokens": [
        {
          "kind": "Colon",
          "position": [
            [
              24,
              1
            ],
            [
              24,
              2
            ]
          ],
          "level": 1
        },
        {
          "kind": {
            "Identifier": "userItems"
          },
          "position": [
            [
              24,
              2
            ],
            [
              24,
              11
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          24,
          1
        ],
        [
          41,
          10
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": {
                "Identifier": "addRecipe"
              },
              "position": [
                [
                  25,
                  5
                ],
                [
                  25,
                  14
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              25,
              5
            ],
            [
              34,
              10
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      26,
                      9
                    ],
                    [
                      26,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "kind"
                  },
                  "position": [
                    [
                      26,
                      10
                    ],
                    [
                      26,
                      14
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      26,
                      14
                    ],
                    [
                      26,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "image"
                  },
                  "position": [
                    [
                      26,
                      15
                    ],
                    [
                      26,
                      20
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "http://path.to.image"
                  },
                  "position": [
                    [
                      26,
                      20
                    ],
                    [
                      28,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  26,
                  9
                ],
                [
                  28,
                  10
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "Coma",
                  "position": [
                    [
                      28,
                      10
                    ],
                    [
                      28,
                      11
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  28,
                  10
                ],
                [
                  28,
                  11
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      29,
                      9
                    ],
                    [
                      29,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      29,
                      10
                    ],
                    [
                      29,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      29,
                      12
                    ],
                    [
                      29,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "en"
                  },
                  "position": [
                    [
                      29,
                      13
                    ],
                    [
                      29,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Add recipe"
                  },
                  "position": [
                    [
                      29,
                      15
                    ],
                    [
                      31,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  29,
                  9
                ],
                [
                  31,
                  10
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      32,
                      9
                    ],
                    [
                      32,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      32,
                      10
                    ],
                    [
                      32,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      32,
                      12
                    ],
                    [
                      32,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ru"
                  },
                  "position": [
                    [
                      32,
                      13
                    ],
                    [
                      32,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Добавить рецепт"
                  },
                  "position": [
                    [
                      32,
                      15
                    ],
                    [
                      34,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  32,
                  9
                ],
                [
                  34,
                  10
                ]
              ],
              "children": []
            }
          ]
        },
        {
          "tokens": [
            {
              "kind": {
                "Identifier": "myAccount"
              },
              "position": [
                [
                  35,
                  5
                ],
                [
                  35,
                  14
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              35,
              5
            ],
            [
              41,
              10
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      36,
                      9
                    ],
                    [
                      36,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      36,
                      10
                    ],
                    [
                      36,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      36,
                      12
                    ],
                    [
                      36,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "en"
                  },
                  "position": [
                    [
                      36,
                      13
                    ],
                    [
                      36,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "My account"
                  },
                  "position": [
                    [
                      36,
                      15
                    ],
                    [
                      38,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  36,
                  9
                ],
                [
                  38,
                  10
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "HashPound",
                  "position": [
                    [
                      39,
                      9
                    ],
                    [
                      39,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ln"
                  },
                  "position": [
                    [
                      39,
                      10
                    ],
                    [
                      39,
                      12
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      39,
                      12
                    ],
                    [
                      39,
                      13
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "ru"
                  },
                  "position": [
                    [
                      39,
                      13
                    ],
                    [
                      39,
                      15
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "ContentBlock": "Мой аккаунт"
                  },
                  "position": [
                    [
                      39,
                      15
                    ],
                    [
                      41,
                      10
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  39,
                  9
                ],
                [
                  41,
                  10
                ]
              ],
              "children": []
            }
          ]
        }
      ]
    },
    {
      "tokens": [
        {
          "kind": "Colon",
          "position": [
            [
              43,
              1
            ],
            [
              43,
              2
            ]
          ],
          "level": 1
        },
        {
          "kind": {
            "Identifier": "searchPlaceholder"
          },
          "position": [
            [
              43,
              2
            ],
            [
              43,
              19
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          43,
          1
        ],
        [
          49,
          6
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": "HashPound",
              "position": [
                [
                  44,
                  5
                ],
                [
                  44,
                  6
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "ln"
              },
              "position": [
                [
                  44,
                  6
                ],
                [
                  44,
                  8
                ]
              ],
              "level": 2
            },
            {
              "kind": "Colon",
              "position": [
                [
                  44,
                  8
                ],
                [
                  44,
                  9
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "en"
              },
              "position": [
                [
                  44,
                  9
                ],
                [
                  44,
                  11
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "ContentBlock": "Find a recipe"
              },
              "position": [
                [
                  44,
                  11
                ],
                [
                  46,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              44,
              5
            ],
            [
              46,
              6
            ]
          ],
          "children": []
        },
        {
          "tokens": [
            {
              "kind": "HashPound",
              "position": [
                [
                  47,
                  5
                ],
                [
                  47,
                  6
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "ln"
              },
              "position": [
                [
                  47,
                  6
                ],
                [
                  47,
                  8
                ]
              ],
              "level": 2
            },
            {
              "kind": "Colon",
              "position": [
                [
                  47,
                  8
                ],
                [
                  47,
                  9
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "ru"
              },
              "position": [
                [
                  47,
                  9
                ],
                [
                  47,
                  11
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "ContentBlock": "Найти рецепт"
              },
              "position": [
                [
                  47,
                  11
                ],
                [
                  49,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              47,
              5
            ],
            [
              49,
              6
            ]
          ],
          "children": []
        }
      ]
    },
    {
      "tokens": [
        {
          "kind": "Colon",
          "position": [
            [
              51,
              1
            ],
            [
              51,
              2
            ]
          ],
          "level": 1
        },
        {
          "kind": {
            "Identifier": "pictureLabel"
          },
          "position": [
            [
              51,
              2
            ],
            [
              51,
              14
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          51,
          1
        ],
        [
          57,
          6
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": "HashPound",
              "position": [
                [
                  52,
                  5
                ],
                [
                  52,
                  6
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "ln"
              },
              "position": [
                [
                  52,
                  6
                ],
                [
                  52,
                  8
                ]
              ],
              "level": 2
            },
            {
              "kind": "Colon",
              "position": [
                [
                  52,
                  8
                ],
                [
                  52,
                  9
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "en"
              },
              "position": [
                [
                  52,
                  9
                ],
                [
                  52,
                  11
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "ContentBlock": "Picture"
              },
              "position": [
                [
                  52,
                  11
                ],
                [
                  54,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              52,
              5
            ],
            [
              54,
              6
            ]
          ],
          "children": []
        },
        {
          "tokens": [
            {
              "kind": "HashPound",
              "position": [
                [
                  55,
                  5
                ],
                [
                  55,
                  6
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "ln"
              },
              "position": [
                [
                  55,
                  6
                ],
                [
                  55,
                  8
                ]
              ],
              "level": 2
            },
            {
              "kind": "Colon",
              "position": [
                [
                  55,
                  8
                ],
                [
                  55,
                  9
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "Identifier": "ru"
              },
              "position": [
                [
                  55,
                  9
                ],
                [
                  55,
                  11
                ]
              ],
              "level": 2
            },
            {
              "kind": {
                "ContentBlock": "Картинка"
              },
              "position": [
                [
                  55,
                  11
                ],
                [
                  57,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              55,
              5
            ],
            [
              57,
              6
            ]
          ],
          "children": []
        }
      ]
    }
  ]
}
